dbms/src/AggregateFunctions/AggregateFunctionGroupBitmap.h

```diff
@@ -35,6 +35,8 @@ class AggregateFunctionBitmap final : public IAggregateFunctionDataHelper<Data,
 
     void merge(AggregateDataPtr place, ConstAggregateDataPtr rhs, Arena *) const override
     {
+        if (!rhs)
```

Comment:
Why `rhs` can be nullptr here?
Why this does not happen for any other aggregate functions?

Reply:
> Why `rhs` can be nullptr here?
> Why this does not happen for any other aggregate functions?

We can not conclude this will not happen for other aggregate functions.  This happened only occasionally using a simple query:
```
    WITH (SELECT mid_seqs FROM cdp_tags WHERE tag_id = 'is_member_1') AS bm
    SELECT tag_id, bitmapCardinality(bitmapAnd(bm, mid_seqs)) FROM cdp_tags;
```
We can only reproduce using a specified table storing bitmap AggregationState data, while if the content of this table is changed, it could not be reproduced. 
And we used GDB to find there's a nullptr here, while after applying this fix, segment fault disappeared.  As a result, it's hard to attach a test.

We have a suspicion that the aggregate function is executed in parallel, and 'merge' is called to combine results from different threads, and in some cases, the `merge` might be triggered given only single working thread, as a result the `rhs` has a nullptr.

Reply:
Can you attach a stack trace of the segfault?
Need to figure out how exactly `rhs` appeared to be nullptr.

Reply:
> Can you attach a stack trace of the segfault?
> Need to figure out how exactly `rhs` appeared to be nullptr.

Sure:
Table:
```
CREATE TABLE cdp_tags (tag_id String, mid_seqs AggregateFunction(groupBitmap, UInt32)) engine=MergeTree() ORDER BY (tag_id) SETTINGS index_granularity=1
```

Query:
```
WITH
    (
        SELECT mid_seqs
        FROM cdp_tags
        WHERE tag_id = 'is_member_1'
    ) AS bm
SELECT
    tag_id,
    bitmapCardinality(bitmapOr(bm, mid_seqs))
FROM cdp_tags
```

```
Thread 41 "ParalInputsProc" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fddc9ffd700 (LWP 40001)]
DB::AggregateFunctionBitmap<unsigned int, DB::AggregateFunctionGroupBitmapData<unsigned int> >::merge (this=0x7fdd8df6eb80, place=0x7fddc81b60a0 "", rhs=0x0)
    at ../dbms/src/AggregateFunctions/AggregateFunctionGroupBitmap.h:38
38	        this->data(place).rbs.merge(this->data(rhs).rbs);
(gdb) bt
#0  DB::AggregateFunctionBitmap<unsigned int, DB::AggregateFunctionGroupBitmapData<unsigned int> >::merge (this=0x7fdd8df6eb80, place=0x7fddc81b60a0 "", rhs=0x0)
    at ../dbms/src/AggregateFunctions/AggregateFunctionGroupBitmap.h:38
#1  0x00007fde6707c711 in DB::FunctionBitmap<DB::BitmapOrImpl, DB::NameBitmapOr>::executeImpl(DB::Block&, std::vector<unsigned long, std::allocator<unsigned long> > const&, unsigned long, unsigned long) () at ../contrib/croaring/roaring/roaring.h:2387
#2  0x00007fde69d45530 in DB::PreparedFunctionImpl::execute (this=0x7fdd8df6e410, block=..., args=..., result=3, input_rows_count=12, dry_run=<optimized out>)
    at ../dbms/src/Functions/IFunction.cpp:473
#3  0x00007fde6a66eab1 in DB::ExpressionAction::execute (this=0x7fdd8dcf0160, block=..., dry_run=false) at /usr/include/c++/9/bits/shared_ptr_base.h:1020
#4  0x00007fde6a6701c5 in DB::ExpressionActions::execute (this=0x7fdd8dfb0010, block=..., dry_run=dry_run@entry=false)
    at ../dbms/src/Interpreters/ExpressionActions.cpp:761
#5  0x00007fde6a501d6a in DB::ExpressionBlockInputStream::readImpl (this=0x7fdd8df89710) at /usr/include/c++/9/bits/shared_ptr_base.h:1020
#6  0x00007fde69dd28f7 in DB::IBlockInputStream::read (this=0x7fdd8df89710) at ../dbms/src/DataStreams/IBlockInputStream.cpp:61
#7  0x00007fde6a501d4b in DB::ExpressionBlockInputStream::readImpl (this=0x7fdd8dfd0f10) at /usr/include/c++/9/bits/shared_ptr_base.h:1020
#8  0x00007fde69dd28f7 in DB::IBlockInputStream::read (this=0x7fdd8dfd0f10) at ../dbms/src/DataStreams/IBlockInputStream.cpp:61
#9  0x00007fde69f49062 in DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::loop (this=this@entry=0x7fdd8dfcd368, thread_num=thread_num@entry=16)
    at /usr/include/c++/9/bits/shared_ptr_base.h:1020
#10 0x00007fde69f49725 in DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::thread (this=0x7fdd8dfcd368, thread_group=..., thread_num=16)
    at /usr/include/c++/9/bits/atomic_base.h:413
#11 0x00007fde69f4a09d in std::__invoke_impl<void, void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::* const&)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>* const&, std::shared_ptr<DB::ThreadGroupStatus> const&, unsigned long const&> (__t=@0x7fddc800b038: 0x7fdd8dfcd368, __f=
    @0x7fddc800b010: (void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::*)(DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler> * const, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long)) 0x7fde69f494e0 <DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::thread(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long)>) at /usr/include/c++/9/ext/atomicity.h:96
#12 std::__invoke<void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::* const&)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>* const&, std::shared_ptr<DB::ThreadGroupStatus> const&, unsigned long const&> (__fn=
    @0x7fddc800b010: (void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::*)(DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler> * const, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long)) 0x7fde69f494e0 <DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::thread(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long)>) at /usr/include/c++/9/bits/invoke.h:95
#13 std::__apply_impl<void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::* const&)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), std:--Type <RET> for more, q to quit, c to continue without paging--
:tuple<DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>*, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long> const&, 0ul, 1ul, 2ul> (__t=...,
    __f=
    @0x7fddc800b010: (void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::*)(DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler> * const, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long)) 0x7fde69f494e0 <DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::thread(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long)>) at /usr/include/c++/9/tuple:1684
#14 std::apply<void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::* const&)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), std::tuple<DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>*, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long> const&> (__t=..., __f=
    @0x7fddc800b010: (void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::*)(DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler> * const, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long)) 0x7fde69f494e0 <DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::thread(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long)>) at /usr/include/c++/9/tuple:1694
#15 ThreadFromGlobalPool::ThreadFromGlobalPool<void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::*)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>*, std::shared_ptr<DB::ThreadGroupStatus>, unsigned long&>(void (DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>::*&&)(std::shared_ptr<DB::ThreadGroupStatus>, unsigned long), DB::ParallelInputsProcessor<DB::UnionBlockInputStream::Handler>*&&, std::shared_ptr<DB::ThreadGroupStatus>&&, unsigned long&)::{lambda()#1}::operator()() const (this=0x7fddc800b000) at ../dbms/src/Common/ThreadPool.h:156
#16 0x00007fde66d27ccc in std::function<void ()>::operator()() const (this=0x7fddc9ff8620) at /usr/include/c++/9/bits/std_function.h:685
#17 ThreadPoolImpl<std::thread>::worker (this=0x7fde6ec0df20 <GlobalThreadPool::instance()::ret>, thread_it=...) at ../dbms/src/Common/ThreadPool.cpp:221
#18 0x00007fde6b86f9c0 in execute_native_thread_routine ()
#19 0x00007fde62a396db in start_thread (arg=0x7fddc9ffd700) at pthread_create.c:463
#20 0x00007fde6255e88f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Reply:
Ok. I suspect that the root cause of this issue is not in `merge` method but in the code that calls it. Because merge method is not supposed to take nullptr instead of pointer to aggregation state.